walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,116 @@
+          return Long.compare(a.getWhen().start(), b.getWhen().start());
+        }
+      };
+
+  /** A comparator for sorting events by their end time in ascending order. */
+  private static final Comparator<Event> ORDER_EVENT_BY_END =
+      new Comparator<Event>() {
+        @Override
+        public int compare(Event a, Event b) {
+          return Long.compare(a.getWhen().end(), b.getWhen().end());
+        }
+      };
+  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Find out who has to be at the meeting
+    neededGuests = request.getAttendees().iterator();
+    // Find out how long they have to be at the meeting
+    long duration = request.getDuration(); 
+    // Set the collection of possible times to be empty
+    result = new ArrayList<TimeRange>(); 
+
+    // window referring to the opening and closing of the window of time where the meeting could be done
+    int windowStart = 0; 
+    int windowClose = 0; 
+
+    Object[] objectArray = events.toArray(); 
+    Event[] eventsOrderedByStart = new Event[objectArray.length]; 
+    Event[] eventsOrderedByEnd = new Event[objectArray.length];
+
+    for (int i = 0; i < events.toArray().length; i++) {
+      eventsOrderedByStart[i] = (Event) objectArray[i];  
+      eventsOrderedByEnd[i] = (Event) objectArray[i]; 
+    }
+
+    Arrays.sort(eventsOrderedByStart, ORDER_EVENT_BY_START);
+    Arrays.sort(eventsOrderedByEnd, ORDER_EVENT_BY_END);
+    int startPointer = 0; 
+    int endPointer = 0; 
+    int nextMeetingEnd = 0; 
+    int nextMeetingStart = 0; 
+    Set<Event> problemEvents = new HashSet<>();
+
+
+
+    // Loop until all meetings have ended
+    while(endPointer < eventsOrderedByEnd.length) {
+    // What comes first, the end of a meeting or a start of one
+      nextMeetingEnd = eventsOrderedByEnd[endPointer].getWhen().end(); 
+      if (startPointer < eventsOrderedByStart.length) {
+      nextMeetingStart = eventsOrderedByStart[startPointer].getWhen().start(); 
+      }
+      if (startPointer == eventsOrderedByStart.length || nextMeetingEnd <= nextMeetingStart) {
+        if (problemEvents.contains(eventsOrderedByEnd[endPointer])) {
+          windowStart = nextMeetingEnd; 
+          problemEvents.remove(eventsOrderedByEnd[endPointer]); 
+        }
+        endPointer++; 
+      } else {
+          if (isEventImportant(eventsOrderedByStart[startPointer], request)) {
+            windowClose = nextMeetingStart; 
+            if (problemEvents.isEmpty()) {
+              if ((windowClose - windowStart) >= duration) {
+                result.add(TimeRange.fromStartEnd(windowStart, windowClose, false)); 
+              }
+            }
+            problemEvents.add(eventsOrderedByStart[startPointer]); 
+          }
+          startPointer++; 
+      }
+  }
+  if (windowStart < (24 * 60) && (24*60 - windowStart) >= duration) {
```

Comment:
You could move 24*60 to a constant. call it TOTAL_MINUTES_PER_DAY or something along those lines.